### ISSUE TITLE
Disappearanced code of main.ts when editing custom.js

### ISSUE THREAD

**Issue by marron9999** [microsoft/pxt-microbit#4896]
**Describe the bug**
Disappearanced code of main.ts when editing custom.js

**To Reproduce**
1. Create new project (name is 'test')
2. Change to editor as javascript 
3. Open explorer, click '+', and create new custom.ts
4. Change to editor as javascript 
5. Write following codes

basic.forever(function () {
    custom.foo(5, "Hello", MyEnum.One)
})

6. Open custom.ts
7. Rename "foo" to "foo2"
8. Change to editor as javascript 

  Change "custom.foo(" to "custom.foo2("

9. Change to editor as blocks

**Expected behavior**
The blocks are displayed correctly

**Screenshots**

Added screenshots as Step7, 8, 9

**micro:bit version (please complete the following information):**

micro:bit v2.0, Interface Version: 0257

**Desktop (please complete the following information):**
 - OS: Windows 10 Home 22H2 build 19045.2364
 - Browser: Chrome  108.0.5359.100（Official Build）(64 bit)

**Additional context**

![step7](https://user-images.githubusercontent.com/46040144/207609053-3548590d-5d31-43ac-a29d-d5e640611484.png)
![step8](https://user-images.githubusercontent.com/46040144/207609055-49a65433-ef84-4f5c-8928-ddee4ba34240.png)
![step9](https://user-images.githubusercontent.com/46040144/207609058-0d73dfa1-8ae6-4bff-8766-2f0114343c91.png)

**Comment by riknoll** [comment]
@marron9999 this is happening because you haven't specified a `blockId` on your function definition.

Every block needs to have a globally unique `blockId`. These blockIds are stored in the XML we save for blocks programs, for example:

```xml
<xml xmlns="http://www.w3.org/1999/xhtml">
    <block type="pxt-on-start">
        <statement name="HANDLER">
            <block type="debugger_keyword">
            </block>
        </statement>
    </block>
</xml>
```

This is a program with 1 block inside of an "on start" event.  The `blockId` for the block inside the "on start" is "debugger_keyword". When MakeCode loads this program, it will search for a defined function that has the debugger_keyword `blockId` and create the appropriate block.

If the `blockId` isn't explicitly set in the comments above a function definition, then it is inferred from the name of the function instead. Changing the name of function will cause that inferred `blockId` to change and MakeCode won't be able to properly load any program that was using the old `blockId`.

tl;dr always specify the `blockId` when defining a custom function. It isn't strictly required, but it prevents issues like these

**Comment by marron9999** [comment]
Thank you.
I understood that there was a way around the problem.

However, I believe the problem in the switching from javascript to blocks.
Because all errors are fixed in javascript  (editor).
I think that part or all of the code disappears by switching the display is a very big problem.

**Comment by martinwork** [comment]
If explicit blockIds are recommended, should the default custom.ts generated by MakeCode should have blockIds?

It appears that there is an automatic process which is getting missed in some cases. If I repeat the repo steps, but save the project as a file instead of switching to blocks, the file can be loaded and switched to blocks.

The same problem occurred when I tried adding explicit blockIds and block= definitions, or subsequently changing a blockId or function name. Blocks mode lost code, but a file saved before the switch to blocks mode could be reloaded and switched to blocks.

Just saving the project to a GitHub repo before the switch to blocks mode seems to make the switch work. 

I noticed while trying the original repro steps, that the console had an error "Uncaught (in promise) Error: TypeError: Cannot read properties of undefined (reading 'replace')" and a list of "duplicate block definition:" warnings.
